File: src/leetcode/java/p1071/Solution.java
package leetcode.java.p1071;
// https://leetcode.com/problems/greatest-common-divisor-of-strings/

public class Solution {
    public String gcdOfStrings(String str1, String str2) {
        if (str1.length() < str2.length())
            return gcdOfStrings(str2, str1);
        String ret = "";
        for (int i = 0; i < str2.length(); ++i) {
            for (int j = i + 1; j <= str2.length(); ++j) {
                if (j - i < ret.length())
                    continue;
                String candidate = str2.substring(i, j);
                if (isDivisible(str1, candidate) && isDivisible(str2, candidate))
                    ret = candidate;
            }
        }
        return ret;
    }

    private boolean isDivisible(String dividend, String divisor) {
        return dividend.equals(divisor.repeat(dividend.length() / divisor.length()));
    }
}

File: src/leetcode/python/p1071/solution.py
"""
https://leetcode.com/problems/greatest-common-divisor-of-strings/description/?envType=study-plan-v2&envId=leetcode-75
"""
class Solution:
    """
    String
    - Time: O(n^2)
    - Space: O(n)
    """
    def gcdOfStrings(self, a: str, b: str) -> str:
        ret= ""
        for i, _ in enumerate(a):
            if i < len(b) and a[i] == b[i]:
                cand= a[:i+1]
                divisible_a= self.divisible(a, cand)
                divisible_b= self.divisible(b, cand)
                if divisible_a and divisible_b:
                    ret= cand
        return ret
    
    def divisible(self, a: str, b: str) -> bool:
        temp= b
        while len(temp) < len(a):
            temp= temp + b
        return temp == a
